Attempt to fix broadcasting bug when insolation is not time-dependent

spencerahill opened the pull request:
Closes #90 

This is a WIP, because c.f. [my comment](https://github.com/brian-rose/climlab/issues/90#issuecomment-474107748), I'm not sure what the actual intended shape is.  But wanted to show the direction I was envisioning going overall to solve this.

spencerahill commented:
OK, thanks to @brian-rose's [comment](https://github.com/brian-rose/climlab/issues/90#issuecomment-474112829), I can confirm that at least for my use case this does lead to the correct shape of `(num_lat, 1)`.

Is there a danger of this dummy dimension name 'x' causing a problem?

Unfortunately I can't run the tests on my local machine with the development code, as there seems to be something going wrong when I build the fortran extensions (on a MacOS 10.14.3).

brian-rose commented:
Probably the same reason why my own development has ground to a halt, see #88.

brian-rose commented:
Looks like you are relying on `xarray.DataArray.expand_dims()` here?

For now at least, it's better to avoid any internal use of xarray functionality because there is absolutely no guarantee that `self.coszen` would be of type `xarray.DataArray`.

Maybe use `numpy.expand_dims` instead?

brian-rose commented:
The `try... except` doesn't break anything, but it might not fix anything either. I suggest adding a basic test just like #90 to `climlab/tests/test_rrtm.py`.

spencerahill commented:
>see #88.

Yes, exactly.  conda-build hanging up indefinitely.

>numpy.expand_dims instead?

Good idea.  Tried that this time around.

>try... except doesn't break anything, but it might not fix anything either. I suggest adding a basic test just like #90 to climlab/tests/test_rrtm.py.

Removed the try/except and added a test..hopefully this does the trick.

brian-rose commented on climlab/tests/test_rrtm.py:
```diff
@@ -60,6 +60,23 @@ def test_multidim():
+    insolation = climlab.solar.insolation.daily_insolation(lat=lat,
+                                                           day=day_of_year)
+    rad = climlab.radiation.RRTMG(name='Radiation', state=state,
+                                  insolation=insolation)
```
change this to 
```
insolation = insolation.values
```
to strip away the xarray metadata (for now).

spencerahill replied:
OK, done in latest commit.

brian-rose commented on climlab/tests/test_rrtm.py:
```diff
@@ -60,6 +60,23 @@ def test_multidim():
+    rad = climlab.radiation.RRTMG(name='Radiation', state=state,
+                                  insolation=insolation)
+    rad.step_forward()
+    assert rad.coszen.shape == rad.Ts.shape
```
This `assert` probably doesn't need to be here. I think in this case `rad.coszen.shape == (num_lat,)` while `rad.Ts.shape == (num_lat, 1)`, so this will fail even if the internal broadcasting is occurring correctly.

spencerahill replied:
OK, done in latest commit.

spencerahill commented:
OK, so my test is failing (e.g. [here](https://travis-ci.org/brian-rose/climlab/jobs/508497780#L1729)), but its resulting in a segfault so I can't discern the exact problem.  But by using my climlab installed via conda for which the build works, I think the problem was that `daily_insolation` was returning a numpy array, and so the `.values` attribute grab was causing an error.

I added a try/except guard in the test to protect against this.  Hopefully that does the trick.

spencerahill commented:
>added a try/except guard in the test to protect against this

Unfortunately that didn't work either.  I'm pretty sure it's not the test logic that's the problem; it's that the fix I've implemented in `rrtmg_sw.py` still isn't working.  E.g. below I've copy-pasted the contents of `test_time_invariant_insolation` into an IPython session using the conda built climlab:

```
IPython 6.1.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import climlab

In [2]: climlab.__path__
Out[2]: ['/Users/shill/Dropbox/miniconda3/lib/python3.5/site-packages/climlab']

In [3]: num_lat = 4
   ...: num_lev = 10
   ...: day_of_year = 80.
   ...: lat = np.linspace(-90., 90., num_lat)
   ...: state = climlab.column_state(num_lev=num_lev, lat=lat)
   ...: insolation = climlab.solar.insolation.daily_insolation(lat=lat,
   ...:                                                        day=day_of_year)
   ...: try:
   ...:     insol_vals = insolation.values
   ...: except AttributeError:
   ...:     insol_vals = insolation
   ...: rad = climlab.radiation.RRTMG(name='Radiation', state=state,
   ...:                               insolation=insol_vals)
   ...: rad.step_forward()
   ...:
Getting ozone data from /Users/shill/Dropbox/miniconda3/lib/python3.5/site-packages/climlab/radiation/data/ozone/apeozone_cam3_5_54.nc
too many axes: 2 (effrank=2), expected rank=1
---------------------------------------------------------------------------
error                                     Traceback (most recent call last)
<ipython-input-3-1640d45e5206> in <module>()
     12 rad = climlab.radiation.RRTMG(name='Radiation', state=state,
     13                               insolation=insol_vals)
---> 14 rad.step_forward()
     15

~/Dropbox/miniconda3/lib/python3.5/site-packages/climlab/process/time_dependent_process.py in step_forward(self)
    325
    326         """
--> 327         tenddict = self.compute()
    328         #  Total tendency is applied as an explicit forward timestep
    329         # (already accounting properly for order of operations in compute() )

~/Dropbox/miniconda3/lib/python3.5/site-packages/climlab/process/time_dependent_process.py in compute(self)
    205         tendencies = {}
    206         ignored = self._compute_type('diagnostic')
--> 207         tendencies['explicit'] = self._compute_type('explicit')
    208         #  Tendencies due to implicit and adjustment processes need to be
    209         #  calculated from a state that is already adjusted after explicit stuff

~/Dropbox/miniconda3/lib/python3.5/site-packages/climlab/process/time_dependent_process.py in _compute_type(self, proctype)
    253             if self.time['steps'] % step_ratio == 0:
    254                 proc.time['active_now'] = True
--> 255                 tenddict = proc.compute()
    256             else:
    257                 # proc.tendencies is unchanged from last subprocess timestep if we didn't recompute it above

~/Dropbox/miniconda3/lib/python3.5/site-packages/climlab/process/time_dependent_process.py in compute(self)
    227                 self.tendencies[varname] += tend
    228         # Finally compute my own tendencies, if any
--> 229         self_tend = self._compute()
    230         #  Adjustment processes _compute method returns absolute adjustment
    231         #  Needs to be converted to rate of change

~/Dropbox/miniconda3/lib/python3.5/site-packages/climlab/process/energy_budget.py in _compute(self)
     70
     71     def _compute(self):
---> 72         tendencies = self._temperature_tendencies()
     73         return tendencies
     74

~/Dropbox/miniconda3/lib/python3.5/site-packages/climlab/process/energy_budget.py in _temperature_tendencies(self)
     58
     59     def _temperature_tendencies(self):
---> 60         self._compute_heating_rates()
     61         tendencies = {}
     62         for varname, value in self.state.items():

~/Dropbox/miniconda3/lib/python3.5/site-packages/climlab/radiation/rrtm/rrtmg_sw.py in _compute_heating_rates(self)
    219                 ciwpmcl, clwpmcl, reicmcl, relqmcl,
    220                 tauaer, ssaaer, asmaer, ecaer,
--> 221                 bndsolvar, indsolvar, solcycfrac)
    222         #  Output is all (ncol,nlay+1) or (ncol,nlay)
    223         self.SW_flux_up = _rrtm_to_climlab(swuflx) + 0.*self.SW_flux_up

error: failed in converting 20th argument `coszen' of _rrtmg_sw.climlab_rrtmg_sw to C/Fortran array
```

Notice that the error is coming at the final, `rad.step_forward()` step, rather than before.  But now I've gotten myself quite confused and need to step away from this for a bit.  One possible next step is, upon returning, try it from one of the Linux clusters I have access to so that I can actually get a properly built dev version on a local machine, so that I can actually see what's going on.

brian-rose commented:
Pretty confused myself. Looks like there is still an issue with the array dimensions or data type for the `coszen` variable that is being passed to the Fortran code. 

Sorry... this "little fix" is not turning out to be so simple.

spencerahill commented:
No worries...that's just how it goes sometime, and better to contribute a fix than to come up with a workaround.  I suspect once #88 is fixed or I can otherwise get a working build of the dev code, I'll be able to figure out the problem pretty easily.

brian-rose commented:
I just realized that the segfaults have nothing to do with the broadcasting of `coszen`. Rather, it's a bad test! The array of latitude grid points should not include -90 and +90. This causes some problems because the assumed shape of the staggered grid has the cell boundaries and cell centers at the same place.

For example, this is a minimal example that triggers a segfault in climlab 0.7.1 (running on linux):
```
import numpy as np
import climlab
num_lat = 4
lat = np.linspace(-90., 90., num_lat)
state = climlab.column_state(lat=lat)
rad = climlab.radiation.RRTMG(state=state)
rad.step_forward()
```

I'll open a separate issue about this... there should be a warning about the invalid grid dimensions.

spencerahill commented:
Superseded by #93.  So closing.
